### ISSUE TITLE
ArangoSearch loses indices after power loss

### ISSUE THREAD

**Issue by thphd** [arangodb/arangodb#13339]
## My Environment

* __ArangoDB Version__:        3.7.5 nightly
* __Storage Engine__:             rocksdb
* __Deployment Mode__:       single
* __Deployment Strategy__:   manual
* __Infrastructure__:               virtualbox windows 10
* __Operating System__:        windows 10
* __Total RAM in your machine__:        4GB
* __Disks in use__:        virtualbox VDI

## Steps to reproduce

1. start ArangoDB in VM, create a View that indexes various attributes on various collections
2. add/modify/delete the documents within those collections continuously
3. shut the VM down in a power-loss manner

__Problem__:

after powering up the VM again, ArangoDB will start just fine with the following output:

```text
C:\<redacted>\ArangoDB3-3.7.5-nightly_win64\usr\bin>arangod --log.level libiresearch=fatal --server.endpoint tcp://0.0.0.0:8529 --rocksdb.max-total-wal-size 10240000 --rocksdb.write-buffer-size 4096000 --rocksdb.max-write-buffer-number 8 --rocksdb.total-write-buffer-size 82000000 --rocksdb.block-cache-size 32000000 --rocksdb.enforce-block-cache-size-limit true --cache.size 10485760 --javascript.v8-contexts 2 --javascript.v8-max-heap 256
2021-01-06T14:52:52Z [4448] INFO [e52b0] ArangoDB 3.7.5-nightly [win64] 64bit, using build heads/3.7-0-g487c64c37d-dirty, VPack 0.1.33, RocksDB 6.8.0, ICU 64.2, V8 7.9.317, OpenSSL 1.1.1h  22 Sep 2020
2021-01-06T14:52:52Z [4448] INFO [43396] {authentication} Jwt secret not specified, generating...
2021-01-06T14:52:52Z [4448] INFO [144fe] using storage engine 'rocksdb'
2021-01-06T14:52:52Z [4448] INFO [3bb7d] {cluster} Starting up with role SINGLE
2021-01-06T14:52:52Z [4448] INFO [6ea38] using endpoint 'http+tcp://0.0.0.0:8529' for non-encrypted requests
2021-01-06T14:52:52Z [4448] INFO [3844e] {authentication} Authentication is turned on (system only)
2021-01-06T14:52:52Z [4448] WARNING [d5c49] {engines} ignoring value for option `--rocksdb.max-write-buffer-number` because it is lower than recommended
2021-01-06T14:52:56Z [4448] ERROR [2885b] {engines} error creating index from definition '{\n  "primarySort" : [\n  ],\n  "storedValues" : [\n  ],\n  "primarySortCompression" : "lz4",\n  "analyzerDefinitions" : [\n    {\n      "name" : "text_zh",\n      "type" : "text",\n      "properties" : {\n        "locale" : "zh.utf-8",\n        "case" : "lower",\n        "stopwords" : [\n        ],\n        "accent" : false,\n        "stemming" : true\n      },\n      "features" : [\n        "position",\n        "norm",\n        "frequency"\n      ]\n    }\n  ],\n  "analyzers" : [\n    "text_zh"\n  ],\n  "fields" : {\n    "title" : {\n    },\n    "content" : {\n    }\n  },\n  "includeAllFields" : false,\n  "trackListPositions" : false,\n  "storeValues" : "none",\n  "id" : "78633652",\n  "type" : "arangosearch",\n  "view" : "h65B15CEF4691/20194085",\n  "objectId" : "78633654"\n}': while checking footer, error: invalid checksum '1771675221'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '78745736'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '78745740'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '78745744'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '78745748'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '78745997'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '78746007'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '78746023'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '78746033'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '79057706'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '79057710'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '79057714'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '79057718'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '79057909'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '79057913'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '79057917'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '79057921'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '79059697'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '79059701'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '79059705'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '79059709'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '79061264'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '79061268'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '79061272'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '79061276'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '79061669'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '79061673'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '79061677'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '79061681'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '79483201'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '79483205'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '79483209'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '79483213'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '79848206'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '79848210'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '79848214'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '79848218'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '81305653'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '81305657'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '81305661'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '81305665'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '81311123'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '81311127'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '81311131'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '81311135'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '81311550'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '81311554'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '81311558'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '81311562'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '81312447'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '81312451'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '81312455'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '81312459'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '81313462'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '81313466'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '81313470'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '81313474'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '81313697'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '81313703'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '81313709'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '81313713'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '81316209'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '81316213'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '81316217'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '81316221'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '81317492'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '81317496'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '81317500'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '81317504'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '81319202'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '81319206'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '81319210'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '81319214'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '81370293'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '81370297'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '81370301'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '81370305'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '81370727'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '81370731'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '81370735'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '81370739'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '81371078'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '81371082'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '81371086'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '81371090'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '81371375'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '81371379'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '81371383'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '81371387'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '81371682'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '81371686'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '81371690'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '81371694'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '81372068'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '81372072'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '81372076'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '81372080'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '81372471'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '81372475'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '81372479'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '81372483'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '81373470'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '81373474'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '81373478'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '81373482'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '81373713'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '81373717'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '81373721'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '81373725'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '81387188'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '81387192'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '81387196'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '81387200'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '81387812'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '81387816'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '81387820'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '81387824'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '81394036'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '81394040'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '81394044'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '81394048'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '81394937'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '81394941'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '81394945'
2021-01-06T14:52:56Z [4448] WARNING [66770] {engines} dropping failed index '81394949'
2021-01-06T14:53:01Z [4448] INFO [cf3f4] ArangoDB (version 3.7.5-nightly [win64]) is ready for business. Have fun!
```

the documents are not lost (as expected); however some/all of the indexed attributes disappear in the previously created arangosearch view. as a consequence, search queries returns nothing.

re-specifying the indexed attributes in the view config solves the problem temporarily. next time the VM shuts down, same thing will happen again.

for example, my view config before power-loss:
```json
  "links": {
    "posts": {
      "analyzers": [
        "text_zh"
      ],
      "fields": {
        "content": {}
      },
      "includeAllFields": false,
      "storeValues": "none",
      "trackListPositions": false
    },
    "users": {
      "analyzers": [
        "text_zh"
      ],
      "fields": {
        "brief": {},
        "name": {}
      },
      "includeAllFields": false,
      "storeValues": "none",
      "trackListPositions": false
    }
  },
```

after power-loss:
```json
  "links": {
    "posts": {
      "analyzers": [
        "text_zh"
      ],
      "fields": {
        "content": {}
      },
      "includeAllFields": false,
      "storeValues": "none",
      "trackListPositions": false
    }
  },
```

in some cases all indexed attributes are gone. and i have to manually add them to the config time after time. 

seems to me that the implementation of ArangoSearch is not very resilient. i think at least the view config should persist between restarts, just like the rest of the documents.

**Comment by thphd** [comment]
some more information: 

apart from power-loss we have a script that `taskkill /f` arangod everyday then restarts it. The goal is to prevent ArangoDB's memory footprint from growing out of control (i have the corresponding command-line options set to minimize memory footprint but it still grows overtime)

if left unattended, on ubuntu could result in OOM kills, on windows7/10 this can cause BSODs.

**Comment by dothebart** [comment]
shouldn't you use `service arangodb3 restart` ?

**Comment by thphd** [comment]
> shouldn't you use `service arangodb3 restart` ?

nope since we're mainly running arangod on windows.
we find that running arangod.exe in a CMD window is more reassuring, comparing to running it as a windows service. when things went wrong i can just look at the window.

**Comment by dothebart** [comment]
but requesting an ordered shutdown (i.e. using curl) is definitively the better way to go: 
https://www.arangodb.com/docs/stable/http/miscellaneous-functions.html#initiate-shutdown-sequence

**Comment by Simran-B** [comment]
If ArangoDB is installed as a service on Windows, then you can stop the service orderly in an elevated command prompt or PowerShell like this:

- `sc stop arangodb` - initiates the service shutdown and terminates immediately. You can run `sc query arangodb` to check the state, which should eventually become `1   STOPPED`
- `net stop arangodb` - this one is blocking, i.e. it waits until the service is shut down before it ends

The service can be started again by using `start` instead of `stop` using either command.

**Comment by gnusi** [comment]
@thphd there is a PR potentially fixing your issue with lost indices
https://github.com/arangodb/arangodb/pull/13313

The fix will become a part of the next minor 3.7 release.

**Comment by maxkernbach** [comment]
Hi @thphd,

Version 3.7.10, which includes the above mentioned fix, has been released in the meantime.

Could you please upgrade and let us know whether the issue persists?
